### PR TITLE
add bench: decimal with byte array and fixed length byte array

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

part of https://github.com/apache/arrow-rs/issues/2388

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: parquet/src/arrow/mod.rs [hunk 1]
```diff
@@ -122,7 +122,7 @@
 experimental!(mod array_reader);
 pub mod arrow_reader;
 pub mod arrow_writer;
-mod buffer;
+experimental!(mod buffer);
```
Comment (tustvold) [inline parquet/src/arrow/mod.rs:125]: Can we avoid this, `experimental`, I really don't want `buffer` exposed externally

File: parquet/benches/arrow_reader.rs [hunk 2]
```diff
@@ -354,6 +431,59 @@ fn create_primitive_array_reader(
...
+        .clone();
+
+    let physical_type = column_desc.physical_type();
+
+    match physical_type {
+        Type::BYTE_ARRAY => {
+            match data_type {
+                DataType::Decimal128(precision, scale) => {
+                    // read decimal data from parquet binary physical type
+                    let convert = DecimalByteArrayConvert::new(
```
Comment (tustvold) [inline parquet/benches/arrow_reader.rs:450]: These functions are being removed...

File: parquet/src/data_type.rs [hunk 3]
```diff
@@ -1235,7 +1241,7 @@ impl FromBytes for ByteArray {
 }
 
 impl FromBytes for FixedLenByteArray {
-    type Buffer = [u8; 8];
+    type Buffer = Vec<u8>;
 
     fn from_le_bytes(bs: Self::Buffer) -> Self {
         Self(ByteArray::from(bs.to_vec()))
```
Comment (tustvold) [inline parquet/src/data_type.rs:1247]: This is performing an unnecessary clone now

**Comment by tustvold** [comment]
I think this will serve as a nice validation of #2528 - the performance should be significantly improved

File: parquet/src/arrow/mod.rs (hunk 1 above)
Comment (liukun4515) [reply parquet/src/arrow/mod.rs:125]: after your refactor https://github.com/apache/arrow-rs/pull/2528, I think we can use some public api like `read_array_reader` to create the reader in the benchmark

File: parquet/src/data_type.rs (hunk 3 above)
Comment (liukun4515) [reply parquet/src/data_type.rs:1247]: good catch

File: parquet/benches/arrow_reader.rs (hunk 2 above)
Comment (liukun4515) [reply parquet/benches/arrow_reader.rs:450]: thanks for your remind, after #2528 merged, and refactor this pr and review it.

When  the #2425 can be reviewed?

Comment (tustvold) [reply parquet/benches/arrow_reader.rs:450]: I need reviews on the PRs it builds on, that's the only thing holding it up being ready for review

**Comment by liukun4515** [comment]
> I think this will serve as a nice validation of #2528 - the performance should be significantly improved

hope so, I will review your refactor work tomorrow.

File: parquet/src/arrow/mod.rs (hunk 1 above)
Comment (liukun4515) [reply parquet/src/arrow/mod.rs:125]: After remove https://github.com/apache/arrow-rs/pull/2529#discussion_r950330507 from your suggestion, we can avoid using the buffer crate in benchmark code

**Comment by liukun4515** [comment]
> I think this will serve as a nice validation of #2528 - the performance should be significantly improved

Hi @tustvold  this pr https://github.com/apache/arrow-rs/pull/2522 will improve the performance?
I can't find the point where to improve the performance in https://github.com/apache/arrow-rs/pull/2522 refactor to remove the `DecimalByteArrayConvert`?

**Comment by liukun4515** [comment]
It's better to remove the complexreader from the array reader @tustvold 
After your refactor, the type of parquet array reader will be matched with the parquet physical type `BOOL,INT32,INT64,INT96,BYTE_ARRAY,FIXED_LENGTH_BYTE_ARRAY`. ❤️

**Comment by tustvold** [comment]
> I can't find the point where to improve the performance in #2522 refactor to remove the DecimalByteArrayConvert?

CompkexObjectArrayReader reads to the row format, i.e. separately boxed ByteArray for each row. ByteArrayReader reads to a contiguous arrow buffer. It also has an optimised path for reading non-nested definition levels. It should therefore be significantly faster, not to mention correctly handling nesting

**Comment by liukun4515** [comment]
great improvement for the decimal reader using the fixed length byte array after the refactor https://github.com/apache/arrow-rs/pull/2528
```
arrow_array_reader/FIXED_LENGTH_BYTE_ARRAY/Decimal128Array/plain encoded, mandatory, no NULLs
                        time:   [453.71 us 462.25 us 472.03 us]
                        change: [-57.944% -57.150% -56.305%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  9 (9.00%) high mild
  1 (1.00%) high severe
arrow_array_reader/FIXED_LENGTH_BYTE_ARRAY/Decimal128Array/plain encoded, optional, no NULLs
                        time:   [438.91 us 442.14 us 445.72 us]
                        change: [-51.955% -51.561% -51.191%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
arrow_array_reader/FIXED_LENGTH_BYTE_ARRAY/Decimal128Array/plain encoded, optional, half NULLs
                        time:   [590.60 us 595.71 us 601.33 us]
                        change: [-31.035% -30.479% -29.836%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
```
@tustvold

**Comment by liukun4515** [comment]
@tustvold  PTAL
just add benchmark for decimal reader using byte array or fixed length byte array.

**Review by tustvold — APPROVED** [review]

**Comment by liukun4515** [comment]
BYTE_ARRAY
```
arrow_array_reader/BYTE_ARRAY/Decimal128Array/plain encoded, mandatory, no NULLs
                        time:   [521.98 us 527.20 us 532.82 us]
                        change: [-64.921% -64.209% -63.473%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 9 outliers among 100 measurements (9.00%)
  7 (7.00%) high mild
  2 (2.00%) high severe
arrow_array_reader/BYTE_ARRAY/Decimal128Array/plain encoded, optional, no NULLs
                        time:   [547.52 us 559.31 us 575.61 us]
                        change: [-60.657% -59.882% -58.990%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  6 (6.00%) high mild
  2 (2.00%) high severe
arrow_array_reader/BYTE_ARRAY/Decimal128Array/plain encoded, optional, half NULLs
                        time:   [637.63 us 643.18 us 649.31 us]
                        change: [-40.139% -39.311% -38.471%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
```

**Pull request merged**